hal/halx86/include/apic.h

```diff
@@ -1,253 +1,301 @@
+#define APIC_EXT2LVTR 0x0520 /* Extended Interrupt 2 Local Vector Table */
+#define APIC_EXT3LVTR 0x0530 /* Extended Interrupt 3 Local Vector Table */
+
+#define MSR_APIC_BASE 0x0000001B
+#define IOAPIC_PHYS_BASE 0xFEC00000
+#define APIC_CLOCK_INDEX 8
+#define ApicLogicalId(Cpu) ((UCHAR)(1<< Cpu))
+
+enum
 {
-    PULONG p;
-
-    p = (PULONG)((ULONG_PTR)APICBase + Offset);
-
-    *p = Value;
-}
-#else
-static __inline VOID APICWrite(ULONG Offset,
-                               ULONG Value)
+    APIC_MT_Fixed = 0,
+    APIC_MT_LowestPriority = 1,
+    APIC_MT_SMI = 2,
+    APIC_MT_RemoteRead = 3,
+    APIC_MT_NMI = 4,
+    APIC_MT_INIT = 5,
+    APIC_MT_Startup = 6,
+    APIC_MT_ExtInt = 7,
+};
+
+enum
 {
-    PULONG p;
-    ULONG CPU = (_APICRead(APIC_ID) & APIC_ID_MASK) >> 24;
-
-    lastregw[CPU] = Offset;
-    lastvalw[CPU] = Value;
-
-    p = (PULONG)((ULONG_PTR)APICBase + Offset);
+    APIC_TGM_Edge,
+    APIC_TGM_Level
+};
 
-    *p = Value;
-}
-#endif
-
-#if 0 
-static __inline ULONG APICRead(ULONG Offset)
+enum
 {
-    PULONG p;
+    APIC_DM_Physical,
+    APIC_DM_Logical
+};
 
-    p = (PULONG)((ULONG_PTR)APICBase + Offset);
-    return *p;
-}
-#else
-static __inline ULONG APICRead(ULONG Offset)
+enum
 {
-    PULONG p;
-    ULONG CPU = (_APICRead(APIC_ID) & APIC_ID_MASK) >> 24;
+    APIC_DSH_Destination,
+    APIC_DSH_Self,
+    APIC_DSH_AllIncludingSelf,
+    APIC_DSH_AllExclusingSelf
+};
 
-    lastregr[CPU] = Offset;
-    lastvalr[CPU] = 0;
+enum
+{
+    APIC_DF_Flat = 0xFFFFFFFF,
+    APIC_DF_Cluster = 0x0FFFFFFF
+};
 
-    p = (PULONG)((ULONG_PTR)APICBase + Offset);
+enum
+{
+    TIMER_DV_DivideBy2 = 0,
+    TIMER_DV_DivideBy4 = 1,
+    TIMER_DV_DivideBy8 = 2,
+    TIMER_DV_DivideBy16 = 3,
+    TIMER_DV_DivideBy32 = 8,
+    TIMER_DV_DivideBy64 = 9,
+    TIMER_DV_DivideBy128 = 10,
+    TIMER_DV_DivideBy1 = 11,
+};
+
+
+typedef union _APIC_BASE_ADRESS_REGISTER
+{
+    ULONG64 Long;
+    struct
+    {
+        ULONG64 Reserved1:8;
+        ULONG64 BootStrapCPUCore:1;
+        ULONG64 Reserved2:2;
+        ULONG64 Enable:1;
+        ULONG64 BaseAddress:40;
+        ULONG64 ReservedMBZ:12;
+    };
+} APIC_BASE_ADRESS_REGISTER;
+
+typedef union _APIC_SPURIOUS_INERRUPT_REGISTER
+{
+    ULONG Long;
+    struct
+    {
+        ULONG Vector:8;
+        ULONG SoftwareEnable:1;
+        ULONG FocusCPUCoreChecking:1;
+        ULONG ReservedMBZ:22;
+    };
+} APIC_SPURIOUS_INERRUPT_REGISTER;
+
+typedef union
+{
+    ULONG Long;
+    struct
+    {
+        ULONG Version:8;
+        ULONG ReservedMBZ:8;
+        ULONG MaxLVT:8;
+        ULONG ReservedMBZ1:7;
+        ULONG ExtRegSpacePresent:1;
+    };
+} APIC_VERSION_REGISTER;
+
+typedef union
+{
+    ULONG Long;
+    struct
+    {
+        ULONG Version:1;
+        ULONG SEOIEnable:1;
+        ULONG ExtApicIdEnable:1;
+        ULONG ReservedMBZ:29;
+    };
+} APIC_EXTENDED_CONTROL_REGISTER;
+
+typedef union _APIC_COMMAND_REGISTER
+{
+    ULONGLONG LongLong;
```

Comment:
For hardware stuff, I'd like the definitions to use types with bit count, this makes things more explicit. `UINT8`, `UINT16`, `UINT32`, `UINT64`
By the way, should we make them packed?